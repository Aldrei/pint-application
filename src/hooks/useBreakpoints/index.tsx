import { useTheme } from '@mui/material/styles';
import useMediaQuery from '@mui/material/useMediaQuery';

export const useBreakpoints = () => {
  const theme = useTheme();

  const goSm = useMediaQuery(theme.breakpoints.up('sm'));
  const goMd = useMediaQuery(theme.breakpoints.up('md'));
  const goLg = useMediaQuery(theme.breakpoints.up('lg'));
  const goXl = useMediaQuery(theme.breakpoints.up('xl'));

  return [goSm, goMd, goLg, goXl];
};
