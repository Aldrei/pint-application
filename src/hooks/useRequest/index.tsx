import * as React from 'react';

import axios from '../../hooks/useConfigAxios';

// import { propertiesVideosService } from '../../services/properties';

import { IServiceRequestStatus } from '../../types';

interface IProps {
  type: 'propertyVideoDelete';
  code: string;
  videoId: string;
}

const useRequest = ({ code, videoId }: IProps) => {
  const [status, setStatus] = React.useState<IServiceRequestStatus>('idle');

  axios.get('https://swapi.dev/api/people', {
    onDownloadProgress: progressEvent => {
      setStatus('loading');
    }
  })
    .then((res) => {
      console.log('DEBUG RESPONSE:', res);
      setStatus('success');
    })
    .catch((err) => {
      console.log('DEBUG ERROR:', err);
      setStatus('failed');
    });

  return { status };
};

export default useRequest;