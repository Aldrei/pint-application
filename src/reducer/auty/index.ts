import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { authService, IAuthServiceAccessTokenRequest, IAuthServiceAccessTokenResponse } from '../../services/auth';

import { IServiceRequest } from '../../types';

export interface IAutyState extends IServiceRequest {
  whoIsAuth: object;
  accessToken: IAuthServiceAccessTokenResponse;
}

const initialState: IAutyState = {
  status: 'idle',
  whoIsAuth: {},
  accessToken: {} as IAuthServiceAccessTokenResponse,
};

// The function below is called a thunk and allows us to perform async logic. It
// can be dispatched like a regular action: `dispatch(incrementAsync(10))`. This
// will call the thunk with the `dispatch` function as the first argument. Async
// code can then be executed and other actions can be dispatched. Thunks are
// typically used to make async requests.
export const authServiceThunk = createAsyncThunk(
  'counter/fetchCount',
  async (data: IAuthServiceAccessTokenRequest) => {
    const response = await authService.accessToken(data);
    // The value we return becomes the `fulfilled` action payload
    return response;
  }
);

export const authSlice = createSlice({
  name: 'auth',
  initialState,
  // The `reducers` field lets us define reducers and generate associated actions
  reducers: {},
  // The `extraReducers` field lets the slice handle actions defined elsewhere,
  // including actions generated by createAsyncThunk or in other slices.
  extraReducers: (builder) => {
    builder
      .addCase(authServiceThunk.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(authServiceThunk.fulfilled, (state, action) => {
        console.log('authServiceThunk fulfilled:', action);
        state.status = 'success';
        state.accessToken = action.payload?.data?.access_token ? action.payload.data : {};
        state.data = action.payload.data;
      })
      .addCase(authServiceThunk.rejected, (state, action) => {
        console.log('authServiceThunk rejected:', action);
        state.status = 'failed';
        state.data = action.payload as object;
        state.accessToken = {};
      });
  },
});

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state: RootState) => state.counter.value)`
// export const selectCount = (state: RootState) => state.counter.value;

export default authSlice.reducer;
