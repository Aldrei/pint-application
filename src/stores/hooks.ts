import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from './index';

// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

export const useSelectorCallback = (state: RootState, reducer: keyof RootState) => state[reducer];
