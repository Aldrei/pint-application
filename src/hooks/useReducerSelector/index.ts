import { RootState, IReducersType } from '../../store';
import { useAppSelector } from '../../store/hooks';

export const useAppSelectorBlaBlaBal = (reducer: keyof IReducersType) => useAppSelector((state: RootState) => state[reducer]);