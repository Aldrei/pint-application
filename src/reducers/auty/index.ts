import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { authService, IAuthServiceAccessTokenRequest, IAuthServiceAccessTokenResponse, IAuthServiceRevokeRequest } from '../../services/auth';
import { AppThunk } from '../../stores';

import { IServiceRequest, IWhoIsAuth } from '../../types';

export interface IAutyState extends IServiceRequest {
  whoIsAuth: IWhoIsAuth;
  accessToken: IAuthServiceAccessTokenResponse;
}

const initialState: IAutyState = {
  name: 'authReducer',
  status: 'idle',
  whoIsAuth: {} as IWhoIsAuth,
  accessToken: {} as IAuthServiceAccessTokenResponse,
};

export const authServiceThunk = createAsyncThunk(
  'auth/access_token',
  async (data: IAuthServiceAccessTokenRequest) => {
    const response = await authService.accessToken(data);
    // The value we return becomes the `fulfilled` action payload

    return response;
  }
);

export const revokeServiceThunk = createAsyncThunk(
  'auth/revoke',
  async (data: IAuthServiceRevokeRequest) => {
    const response = await authService.revoke(data);
    // The value we return becomes the `fulfilled` action payload

    return response;
  }
);

export const whoServiceThunk = createAsyncThunk(
  'auth/who_is_auth',
  async () => {
    const response = await authService.who();
    // The value we return becomes the `fulfilled` action payload

    return response;
  }
);

export const authSlice = createSlice({
  name: 'auth',
  initialState,
  // The `reducers` field lets us define reducers and generate associated actions
  reducers: {
    clearAccessToken: (state) => {
      state.accessToken = {} as IAuthServiceAccessTokenResponse;
      state.whoIsAuth = {} as IWhoIsAuth;
    }
  },
  // The `extraReducers` field lets the slice handle actions defined elsewhere,
  // including actions generated by createAsyncThunk or in other slices.
  extraReducers: (builder) => {
    builder
      /** Access token */
      .addCase(authServiceThunk.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(authServiceThunk.fulfilled, (state, action) => {
        state.status = 'success';
        state.accessToken = action.payload?.data?.access_token ? action.payload.data : {};
        state.data = action.payload.data;
      })
      .addCase(authServiceThunk.rejected, (state, action) => {
        state.status = 'failed';
        state.data = action.payload as object;
        state.accessToken = {};
      })
      /** Revoke. */
      .addCase(revokeServiceThunk.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(revokeServiceThunk.fulfilled, (state, action) => {
        state.status = 'success';
        state.data = action.payload.data;
        state.accessToken = {};
        state.whoIsAuth = {} as IWhoIsAuth;
      })
      .addCase(revokeServiceThunk.rejected, (state, action) => {
        state.status = 'failed';
        state.data = action.payload as object;
      })
      /** Who is auth. */
      .addCase(whoServiceThunk.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(whoServiceThunk.fulfilled, (state, action) => {
        state.status = 'success';
        state.whoIsAuth = action.payload.data;
      })
      .addCase(whoServiceThunk.rejected, (state, action) => {
        state.status = 'failed';
        state.whoIsAuth = action.payload as IWhoIsAuth;
      });
  },
});

const { clearAccessToken } = authSlice.actions;

export const clearAccessTokenReducer = (): AppThunk => (dispatch) => {
  dispatch(clearAccessToken());
};

export default authSlice.reducer;
