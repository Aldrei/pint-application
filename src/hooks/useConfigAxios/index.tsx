import axios from 'axios';

import { clearAccessTokenReducer } from '../../reducers/auty';
import { useAppDispatch } from '../../stores/hooks';

const api = axios.create({
  baseURL: process.env.REACT_APP_API_BASE_URL,
});

const getToken = () => {
  try {
    const persistImob = localStorage.getItem('persist:imob');
    if (persistImob) {
      const persistImobData = JSON.parse(persistImob);

      if (persistImobData.authReducer) {
        const authReducerData = Object.assign({}, JSON.parse(persistImobData.authReducer));
        return authReducerData?.accessToken?.access_token;
      }
    }

    return '';
  } catch (error) {
    console.log(error);
    return '';
  }
};


// Add a request interceptor
api.interceptors.request.use(function (config) {
  if (config.headers) config.headers['Authorization'] = `Bearer ${getToken()}`;
  return config;
}, function (error) {
  // Do something with request error
  return Promise.reject(error);
});

export const useConfigAxios = () => {
  const dispatch = useAppDispatch();

  // Add a response interceptor
  api.interceptors.response.use((response) => {
    console.log('api.interceptors.response success...');
    // Any status code that lie within the range of 2xx cause this function to trigger
    /** NOTE: Suppress 'A non-serializable value was detected in the state, in the path' error */
    return JSON.parse(JSON.stringify(response));
  }, (error) => {
    console.log('api.interceptors.response error...', { error });
    // Any status codes that falls outside the range of 2xx cause this function to trigger
    // return Promise.reject(error);

    if (error.response.status === 401) {
      dispatch(clearAccessTokenReducer());
    }

    /** NOTE: Suppress 'A non-serializable value was detected in the state, in the path' error */
    return JSON.parse(JSON.stringify(error.response));
  });

  return [];
};

export default api;