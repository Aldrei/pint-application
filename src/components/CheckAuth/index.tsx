import React, { useEffect, useState } from 'react';

import { Navigate } from 'react-router-dom';

import { IAutyState, whoServiceThunk } from '../../reducers/auty';
import { useAppSelectorBlaBlaBal } from '../../hooks/useReducerSelector';
import { useAppDispatch } from '../../hooks/useReducerDispatch';

import { useConfigAxios } from '../../hooks/useConfigAxios';

import Menu from '../Menu';
import Header from '../Header';

const CheckAuth = ({ children }: { children: JSX.Element }) => {
  const dispatch = useAppDispatch();

  const { accessToken } = useAppSelectorBlaBlaBal('authReducer') as IAutyState;
  const [menuActive, setMenuActive] = useState(false);

  useConfigAxios();

  useEffect(() => {
    /**
     * Soon after user loggin and is redirect to dashboard page, getToke() don't can get the new token from localStorage.
     * So, pass the new token by parameter.
    */
    if (accessToken?.access_token) dispatch(whoServiceThunk(accessToken?.access_token));
  }, [accessToken]);

  const toggleDrawer =
    (
      // event: React.KeyboardEvent | 
      event: React.MouseEvent, open: boolean) => {
      
      // if (
      //   event.type === 'keydown' &&
      //   ((event as React.KeyboardEvent).key === 'Tab' ||
      //     (event as React.KeyboardEvent).key === 'Shift')
      // ) {
      //   return;
      // }

      setMenuActive(open);
    };

  if (!accessToken?.access_token) 
    return <Navigate to="/login" />;

  return (
    <React.Fragment>
      <Header toggleMenu={(e, action) => toggleDrawer(e, action)} />
      <Menu
        handleClose={() => {
          setMenuActive(false);
        }}
        menuActive={menuActive} />
      {children}
    </React.Fragment>
  );
};

export default CheckAuth;
