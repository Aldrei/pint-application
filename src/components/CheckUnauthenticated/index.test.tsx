import { render } from '@testing-library/react';
import React from 'react';

import { useAppSelectorBlaBlaBal } from '../../hooks/useReducerSelector';
import { IAutyState } from '../../reducers/auty';

import CheckAuth from './index';
import { useLocation } from 'react-router-dom';

jest.mock('../../hooks/useReducerSelector', () => ({
  useAppSelectorBlaBlaBal: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  Navigate: jest.fn(),
  useLocation: jest.fn(),
}));

jest.mock('../../constants/routes', () => ({
  ROUTES: {
    login: {
      path: '/login'
    }
  }
}));

describe('CheckUnauthenticated component', () => {
  const useAppSelectorBlaBlaBalMocked = useAppSelectorBlaBlaBal as jest.MockedFunction<typeof useAppSelectorBlaBlaBal>;
  const useLocationMocked = useLocation as jest.MockedFunction<typeof useLocation>;

  beforeEach(() => {
    jest.resetAllMocks();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('Should render children correctly', () => {
    useAppSelectorBlaBlaBalMocked.mockReturnValue({ accessToken: { access_token: 'dummy-token' } } as IAutyState);
    useLocationMocked.mockReturnValue({
      pathname: '/test', 
      search: '', 
      hash: '',
      state: undefined,
      key: ''
    });

    const nodeEl = render(<CheckAuth><div>Children element</div></CheckAuth>);
    expect(nodeEl.baseElement).toHaveTextContent('Children element');
    expect(nodeEl.baseElement).toMatchSnapshot();
  });

  it('Should redirect to Login page', () => {
    useAppSelectorBlaBlaBalMocked.mockReturnValue({ accessToken: { access_token: 'dummy-token' } } as IAutyState);
    useLocationMocked.mockReturnValue({
      pathname: '/login', 
      search: '', 
      hash: '',
      state: undefined,
      key: ''
    });

    const nodeEl = render(<CheckAuth><div>Children element</div></CheckAuth>);
    expect(nodeEl.baseElement).not.toHaveTextContent('Children element');
    expect(nodeEl.baseElement).toMatchSnapshot();
  });
});
