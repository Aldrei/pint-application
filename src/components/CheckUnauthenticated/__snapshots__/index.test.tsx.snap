// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckUnauthenticated component Should redirect to Login page 1`] = `
<body>
  <div />
</body>
`;

exports[`CheckUnauthenticated component Should render children correctly 1`] = `
<body>
  <div>
    <div>
      Children element
    </div>
  </div>
</body>
`;
